pyhelper_utils/general.py

```diff
@@ -84,6 +84,42 @@ def inner(*args: Any, **kwargs: Any) -> Any:
     return wrapper
 
 
+def retry_on_exception(
+    retry: int = 0,
+    retry_interval: int = 1,
+    logger: Logger | None = None,
+) -> Any:
+    """
+    Decorator to retry a function on exceptions until the retry limit is reached.
+
+    Args:
+        retry (int): Number of retries to attempt (excluding initial attempt) before throwing an exception.
+        retry_interval (int): Number of seconds to wait between retries.
+        logger (Logger): logger to use, if not passed no logs will be displayed.
+
+    Returns:
+        any: the underline function return value.
+    """
+
+    def wrapper(func: Callable) -> Callable:
+        @wraps(func)
+        def inner(*args: Any, **kwargs: Any) -> Any:
+            for idx in range(retry + 1):
```

Comment:
```suggestion
            for idx in range(retry, retry + 1):
```

```
In [25]: retry = 1

In [26]: for idx in range(retry + 1): print(idx)
0
1

In [27]: for idx in range(1, retry + 1): print(idx)
1

```

Reply:
2 is ok, if you asked for 1 retry then the first call is the not retry call, any call after this count as retry.

Reply:
maybe we need to combine the two function?

Also it will be more readable if on first try we call the function outside the for loop.

Reply:
also we may need to add `delay` for retry

Reply:
and exception type (default to `Exception`)

Reply:
https://github.com/RedHatQE/pyhelper-utils/pull/46/files#r1590273410 agree

Reply:
@rnetser @myakove I'll be OOO until this Thursday, I can work on combining the two functions when I am back. If you end up doing it before then, just let me know and I'll close this PR.